reactor-core/src/main/java/reactor/core/publisher/SinkManyBestEffort.java

```diff
@@ -0,0 +1,385 @@
+	}
+
+	@Override
+	public void emitComplete() {
+		//no particular error condition handling for onComplete
+		@SuppressWarnings("unused")
+		Emission emission = tryEmitComplete();
+	}
+
+	@Override
+	public void emitError(Throwable error) {
+		Emission result = tryEmitError(error);
+		if (result == Emission.FAIL_TERMINATED) {
+			Operators.onErrorDroppedMulticast(error, subscribers);
+		}
+	}
+
+	@Override
+	public void emitNext(T value) {
+		switch(tryEmitNext(value)) {
+			case FAIL_ZERO_SUBSCRIBER:
+				//we want to "discard" without rendering the sink terminated.
+				// effectively NO-OP cause there's no subscriber, so no context :(
+				break;
+			case FAIL_OVERFLOW:
+				Operators.onDiscard(value, currentContext());
+				//the emitError will onErrorDropped if already terminated
+				emitError(Exceptions.failWithOverflow("Backpressure overflow during Sinks.Many#emitNext"));
+				break;
+			case FAIL_CANCELLED:
+				Operators.onDiscard(value, currentContext());
+				break;
+			case FAIL_TERMINATED:
+				Operators.onNextDroppedMulticast(value, subscribers);
+				break;
+			case OK:
+				break;
+			case FAIL_NON_SERIALIZED:
+				throw new IllegalStateException("Unexpected return code");
+		}
+	}
+
+	@Override
+	public int currentSubscriberCount() {
+		return subscribers.length;
+	}
+
+	@Override
+	public Flux<T> asFlux() {
+		return this;
+	}
+
+	@Override
+	public void subscribe(CoreSubscriber<? super T> actual) {
+		Objects.requireNonNull(actual, "subscribe(null) is forbidden");
+
+		Inner<T> p = new Inner<>(actual, this);
+		actual.onSubscribe(p);
+
+		if (p.isCancelled()) {
+			return;
+		}
+
+		if (add(p)) {
+			if (p.isCancelled()) {
+				remove(p);
+			}
+		}
+		else {
+			Throwable e = error;
+			if (e != null) {
+				actual.onError(e);
+			}
+			else {
+				actual.onComplete();
+			}
+		}
+	}
+
+	boolean add(Inner<T> s) {
+		Inner<T>[] a = subscribers;
+		if (a == TERMINATED) {
+			return false;
+		}
+
+		synchronized (this) {
+			a = subscribers;
+			if (a == TERMINATED) {
+				return false;
+			}
+			int len = a.length;
+
+			@SuppressWarnings("unchecked")
+			Inner<T>[] b = new Inner[len + 1];
+			System.arraycopy(a, 0, b, 0, len);
+			b[len] = s;
+
+			subscribers = b;
+
+			return true;
+		}
+	}
+
+	@SuppressWarnings("unchecked")
+	void remove(Inner<T> s) {
+		Inner<T>[] a = subscribers;
+		if (a == TERMINATED || a == EMPTY) {
+			return;
+		}
+
+		synchronized (this) {
+			a = subscribers;
+			if (a == TERMINATED || a == EMPTY) {
+				return;
+			}
+			int len = a.length;
+
+			int j = -1;
+
+			for (int i = 0; i < len; i++) {
+				if (a[i] == s) {
+					j = i;
+					break;
+				}
+			}
+			if (j < 0) {
+				return;
+			}
+			if (len == 1) {
+				subscribers = EMPTY;
+				return;
+			}
+
+			Inner<T>[] b = new Inner[len - 1];
+			System.arraycopy(a, 0, b, 0, j);
+			System.arraycopy(a, j + 1, b, j, len - j - 1);
+
+			subscribers = b;
+		}
+	}
+
+	static class Inner<T> extends AtomicBoolean implements InnerProducer<T> {
```

Comment:
can't we reuse at least the inner class between this Sink and the existing processor?

Reply:
not really, since the `DirectInner` terminates its downstream on backpressure. (also it would imply a common interface for `DirectProcessor` and `SinkManyBestEffort` for the inners to access `parent.remove` method)

Reply:
But is this common interface a big deal?

The termination behaviour can be changed with two methods - emitXxx and tryEmitXxx, so that BestEffort will call "tryEmitXxx", while DirectProcessor would stick to emitXxx that calls tryEmitXxx and terminates

Reply:
I've opened draft PR https://github.com/reactor/reactor-core/pull/2393 because I'm still not convinced this is in that much of a better state after trying your suggestion